### DISCUSSION TITLE
Variable vs Constraint-based solver

### DISCUSSION THREAD

**Discussion by perdasilva** [operator-framework/deppy#34] [General]
Hey everyone,

I wanted to ask you what you think about the work I've put together in [this](https://github.com/operator-framework/deppy/pull/33) PR.

Currently, the solver takes as [input](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/solve.go#L132) a slice of [variables](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/variable.go#L19). `Variable` is an interface that connects an [Identifier](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/variable.go#L5) with a set of [Constraints](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/constraints.go#L13) for the purposes of generating `gini` literals when the constraint is [applied](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/constraints.go#L15) (e.g. see the [dependency constraint](https://github.com/operator-framework/deppy/blob/90318e1cd652d4b5b44b63c94293264be979a98f/internal/solver/constraints.go#L117)).

I've refactored the `Solver` to take as input a slice of `Constraint`s, which were themselves refactored to contain the `subject` of the constraint ([i.e.](https://github.com/operator-framework/deppy/pull/33/files#diff-820c669d24ab002f8a57cdcd725324467be25fc300597fe5466cbbfb8875bee5R16)). I did this because I *think* it will make it easier to express arbitrary constraints (inc. conjunctions and disjunctions of arbitrary constraints). 

For instance, if we have two variables:

A:
- mandatory
- dependency(X, Y, Z)
B:
- mandatory
- dependency(C, D, E)

It would be hard to express something like `(A OR B)`, especially given that each `Variable` [must only appear once in the input ](https://github.com/operator-framework/deppy/blob/main/internal/solver/lit_mapping.go#L53).

With a constraint based approach, you could simply add a new constraint, e.g. `Or(Mandatory("a"), Mandatory("b"))`.

**Comment by akihikokuroda** [comment]
Is [this](https://github.com/operator-framework/deppy/blob/5c31d0bbe5cfbdfcfca2065cac5747cd27df2ca2/internal/solver/solve_test.go#L160) **a OR b** (c depends on)?

**Reply by perdasilva** [reply]
That's right. The dependency constraint is an implication: A -> B, which gets translated to `(B or ~A)`. 
Because a package dependency can be fulfilled by multiple bundles, you'd have A -> B1 or A-> B2 or A -> B3, etc.
This simplifies to `(B1 or B2 or B3 or .. Bn or ~A)`. If we want to install A, for the clause to the `true`, one (or more) of B1..Bn must be true.
This is why we also need another constraint over the B1..Bn to say that `AtMost` 1 can be true.

**Comment by perdasilva** [comment]
After an interesting [conversation](https://kubernetes.slack.com/archives/C045L6WK27N/p1665484881361889) on slack. It seems going the `Variable` way, might be the right way.
